plugins/sources/telegraf/provider.go

```diff
@@ -0,0 +1,206 @@
+package telegraf
+
+import (
+	"fmt"
+	"net/url"
+	"os"
+	"reflect"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/influxdata/telegraf/plugins/inputs/procstat"
+
+	"github.com/golang/glog"
+
+	"github.com/influxdata/telegraf"
+	telegrafInputs "github.com/influxdata/telegraf/plugins/inputs"
+	wf "github.com/wavefronthq/wavefront-kubernetes-collector/internal/metrics"
+)
+
+// NewProvider creates a Telegraf source
+func NewProvider(uri *url.URL) (wf.MetricsSourceProvider, error) {
+
+	for _, pair := range os.Environ() {
+		glog.Infof("env: %v", pair)
+	}
+
+	glog.Infof("[telegraf.NewProvider] - inputs: %v -------------", telegrafInputs.Inputs)
+
+	var sources []wf.MetricsSource
+	for name, creator := range telegrafInputs.Inputs {
+		plugin := creator()
+		if name == "procstat" {
+			continue // I hate me ;-)
+		}
+		sources = append(sources, newTelegrafPluginSource(name+" plugin", plugin))
+	}
+
+	for _, exe := range []string{"wavefront-collector", "dockerd"} {
+		procstat := telegrafInputs.Inputs["procstat"]().(*procstat.Procstat)
+		procstat.Exe = exe
+		procstat.PidFinder = "native"
+		sources = append(sources, newTelegrafPluginSource("procstat '"+exe+"' plugin", procstat))
+	}
+
+	return &telegrafProvider{sources: sources}, nil
+}
+
+type telegrafProvider struct {
+	sources []wf.MetricsSource
+}
+
+func (p telegrafProvider) GetMetricsSources() []wf.MetricsSource {
+	return p.sources
+}
+
+func (p telegrafProvider) Name() string {
+	return "Telegraf Source"
+}
+
+type telegrafPluginSource struct {
+	name   string
+	source string
+	plugin telegraf.Input
+	points []*wf.MetricPoint
+	mux    sync.Mutex
+}
+
+func newTelegrafPluginSource(name string, plugin telegraf.Input) *telegrafPluginSource {
+	hostname := os.Getenv("POD_NODE_NAME")
+	glog.Infof("hostname: '%s'", hostname)
+
+	tsp := &telegrafPluginSource{name: name, plugin: plugin, source: hostname}
+	ticker := time.NewTicker(5 * time.Second)
+	go func() {
+		for {
+			select {
+			case <-ticker.C:
+				err := plugin.Gather(tsp)
```

Comment:
We'd want to do the actual gathering within the `ScrapeMetrics` function at the intervals configured on the collector level.